src/Data/Voting.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE KindSignatures #-}
+module Data.Voting
+where
+
+import Prelude       hiding ((.))
+import Data.Function hiding ((.))
+import Control.Applicative hiding (empty)
+import Control.Category
+
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe
+import Data.Set (Set)
+import qualified Data.Set as Set
+
+data VoterKind = VoterKind
+
+data Voter = Voter Int
+  deriving (Eq, Ord, Show)
+
+data Vote = Yes | No
+
+data Delegation v (a :: VoterKind) (b :: VoterKind)
+    = Delegation v v
+    | IdentityDelegation
+    | NoDelegation
+
+identity :: Voter -> Delegation Voter 'VoterKind 'VoterKind
+identity v = Delegation v v
+
+delegate :: Voter -> Voter -> Delegation Voter 'VoterKind 'VoterKind
+delegate from to = Delegation from to
+
+instance Eq v => Category (Delegation v) where
+    id = IdentityDelegation
+    NoDelegation . x = NoDelegation
+    x . NoDelegation = NoDelegation
+    IdentityDelegation . (Delegation a b) = Delegation a b
+    (Delegation a b) . IdentityDelegation = Delegation a b
+    (Delegation b1 c) . (Delegation a b0)
+      | b0 == b1  = Delegation a c
+      | otherwise = NoDelegation
```

Comment:
so `NoDelegation` is the miss of all partial functions here.  (not really a comment just thinking out loud.)